File: FE/pp_ast.py
#!/bin/env python3
"""Pretty printer for AST using unicode box chars:

https://www.w3.org/TR/xml-entity-names/025.html
"""

import logging

from typing import Any

from FE import cwast
from FE import symbolize
from FE import type_corpus
from FE import typify
from FE import canonicalize
from FE import eval

logger = logging.getLogger(__name__)

_INDENTATION = 4


def _EmitLine(line: list[str], indent: int, fout, active_columns: list[int], is_last: bool):
    if indent >= 0:
        spaces = [" "] * (indent * _INDENTATION + 1)
        for a in active_columns:
            spaces[a * _INDENTATION] = "┃"
        spaces[indent * _INDENTATION] = "┗" if is_last else "┣"

        print("".join(spaces), file=fout, end="")
    print(" ".join(line), file=fout)


def _DumpList(name: str, lst, indent,  labels: dict[Any, str], fout, active_columns, is_last):
    if lst:

        _EmitLine([name], indent + 1,
                  fout, active_columns,  is_last)
        active_columns.append(indent + 2)
        for n in lst:
            if n == lst[-1]:
                active_columns.pop(-1)

            _DumpNode(n, indent + 2, labels, fout,
                      active_columns, n == lst[-1])

    else:
        _EmitLine([name, "[Empty]"],
                  indent + 1, fout, active_columns, is_last)


def _GetFlags(node) -> str:
    cls = type(node)
    flags = []
    has_flags = False
    for nfd in cls.ATTRS:
        if nfd.name == 'doc':
            continue
        assert nfd.kind == cwast.NFK.ATTR_BOOL, f"{nfd.name}"
        has_flags = True
        val = getattr(node, nfd.name)
        if val:
            flags.append(nfd.name)
    if has_flags:
        return f"[{','.join(sorted(flags))}]"
    return ""


def _DumpNode(node: Any, indent: int,  labels: dict[Any, str],  fout, active_columns, is_last):
    cls = type(node)
    line = [cls.__name__]
    flags = _GetFlags(node)
    if flags:
        line.append(flags)
    for nfd in cls.KIND_FIELDS:
        # these are all enum
        val = getattr(node, nfd.name).name
        line.append(val)
    for nfd in cls.STR_FIELDS:
        val = getattr(node, nfd.name)
        if val is not None:
            if nfd.kind == cwast.NFK.NAME:
                val = val.name
                assert val is not None
            if nfd.name == "name":
                line.append(val)
            else:
                line.append(f"{nfd.name}={val}")
    label = labels.get(node)
    if label is not None:
        line.append(f"label={label}")

    for name in cls.X_FIELD_NAMES:
        val = getattr(node, name)
        if val is None:
            continue
        if name == "x_srcloc":
            continue
        if name == "x_type":
            if isinstance(node, cwast.TypeAuto):
                val = val.name
            else:
                continue
        if name == "x_eval":
            continue
        if name == "x_symbol" or name == "x_target" or name == "x_poly_mod":
            def_sym = getattr(node, name)
            assert def_sym in labels, f"{node} -> {def_sym}"
            line.append(f"{name}={labels[def_sym]}")
            continue
        if name == "x_import" or name == "x_symtab":
            continue
        line.append(f"{name}={val}")

    _EmitLine(line, indent, fout, active_columns, is_last)

    if cls.NODE_FIELDS:
        active_columns.append(indent + 1)

        last = cls.NODE_FIELDS[-1]

        for nfd in cls.NODE_FIELDS:
            if nfd == last:
                active_columns.pop(-1)
            if nfd.kind == cwast.NFK.NODE:
                _DumpNode(getattr(node, nfd.name),
                          indent + 1, labels, fout, active_columns, nfd == last)
            else:
                assert nfd.kind == cwast.NFK.LIST
                _DumpList(nfd.name, getattr(node, nfd.name),
                          indent, labels, fout, active_columns, nfd == last)


_LABEL_KIND = {
    cwast.DefMod: "m",
    cwast.DefFun: "f",
    cwast.DefRec: "r",
    cwast.RecField: "F",
    cwast.DefEnum: "e",
    cwast.DefMacro: "x",
    cwast.DefType: "t",
    cwast.DefGlobal: "g",
    cwast.DefVar: "v",
    cwast.FunParam: "p",
}


def _LabeDefs(node: Any, prefix: list[int], out: dict[Any, str]):
    tag = _LABEL_KIND.get(type(node))
    if tag:
        out[node] = tag + ".".join((str(x) for x in prefix))
    for nfd in type(node).NODE_FIELDS:
        if nfd.kind == cwast.NFK.LIST:
            prefix.append(-1)
            for n, child in enumerate(getattr(node, nfd.name)):
                prefix[-1] = n
                _LabeDefs(child, prefix, out)
            prefix.pop(-1)


def _ExtractSymDefLabels(mods: list[cwast.DefMod], out: dict[Any, str]):
    for n, mod in enumerate(mods):
        _LabeDefs(mod, [n], out)


def _LabelTargets(node: Any, n: int, out: dict[Any, str]):
    block_counter = 0
    expr_counter = 0

    def visit(node):
        nonlocal n, out, block_counter, expr_counter
        if isinstance(node, cwast.StmtBlock):
            out[node] = f"b{n}.{block_counter}"
            block_counter += 1
        elif isinstance(node, cwast.ExprStmt):
            out[node] = f"s{n}.{expr_counter}"
            expr_counter += 1

    cwast.VisitAstRecursively(node, visit)


def _ExtractTargetLabels(mods: list[cwast.DefMod], out: dict[Any, str]):
    for n, mod in enumerate(mods):
        _LabelTargets(mod, n, out)


def DumpMods(mods: list[cwast.DefMod], fout):
    labels: dict[Any, str] = {}
    _ExtractSymDefLabels(mods, labels)

    _ExtractTargetLabels(mods, labels)

    for mod in mods:
        print("", file=fout)
        _DumpNode(mod, -1, labels, fout, [], True)

############################################################
#
############################################################


def main(argv: list[str]):
    cwast.ASSERT_AFTER_ERROR = False
    assert len(argv) == 1
    fn = argv[0]
    fn, ext = os.path.splitext(fn)
    assert ext in (".cw", ".cws")

    cwd = os.getcwd()
    main = str(pathlib.Path(fn).resolve())
    mp = mod_pool.ReadModulesRecursively(pathlib.Path(
        cwd) / "Lib", [main], add_builtin=fn != "Lib/builtin")
    for mod in mp.mods_in_topo_order:
        canonicalize.FunRemoveParentheses(mod)
    macro.ExpandMacrosAndMacroLike(mp.mods_in_topo_order)
    symbolize.SetTargetFields(mp.mods_in_topo_order)
    symbolize.ResolveSymbolsInsideFunctions(
        mp.mods_in_topo_order, mp.builtin_symtab)
    for mod in mp.mods_in_topo_order:
        symbolize.VerifySymbols(mod)

    tc = type_corpus.TypeCorpus(type_corpus.STD_TARGET_X64)
    typify.AddTypesToAst(mp.mods_in_topo_order, tc)
    for mod in mp.mods_in_topo_order:
        typify.VerifyTypesRecursively(mod, tc, typify.VERIFIERS_WEAK)

    eval.DecorateASTWithPartialEvaluation(mp.mods_in_topo_order)
    for mod in mp.mods_in_topo_order:
        eval.VerifyASTEvalsRecursively(mod)

    DumpMods(mp.mods_in_topo_order, sys.stdout)


if __name__ == "__main__":
    import sys
    import os
    import pathlib
    from FE import mod_pool
    from FE import macro

    logging.basicConfig(level=logging.WARN)
    logger.setLevel(logging.WARN)
    main(sys.argv[1:])

File: FE/pp_ast.cc
#include "FE/pp_ast.h"

#include <fstream>
#include <functional>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

#include "FE/cwast_gen.h"
#include "FE/eval.h"
#include "FE/type_corpus.h"
#include "Util/assert.h"

namespace cwerg::fe {

std::string MakeLabel(std::string_view c,
                      const std::vector<std::string_view>& prefix) {
  std::string out;
  // assume each prefix element is size 1
  out.reserve(prefix.size() * 2 + c.size());
  std::string_view sep = c;
  for (const auto& p : prefix) {
    out += sep;
    sep = ".";
    out += p;
  }
  return out;
}

// Helper for label generation
class Prefix {
  std::vector<int> prefix_;
  // Avoids creating the same strings over and over again
  std::vector<std::string> digits_cache_;
  std::string tmp_;

 public:
  Prefix() {}

  void PushLevel() { prefix_.push_back(-1); }

  void PopLevel() { prefix_.pop_back(); }

  void SetCurrent(int i) { prefix_.back() = i; }

  std::string GetLabel(std::string_view c) {
    tmp_.clear();
    std::string_view sep = c;
    for (const auto& p : prefix_) {
      while (digits_cache_.size() <= p) {
        digits_cache_.push_back(std::to_string(digits_cache_.size()));
      }
      tmp_ += sep;
      sep = ".";
      tmp_ += digits_cache_[p];
    }
    return tmp_;
  }
};

std::string_view GetLabelTag(NT kind) {
  switch (kind) {
    case NT::DefMod:
      return "M";
    case NT::DefFun:
      return "F";
    case NT::DefRec:
      return "R";
    case NT::DefEnum:
      return "E";
    case NT::EnumVal:
      return "e";
    case NT::DefMacro:
      return "X";
    case NT::DefType:
      return "T";
    case NT::DefGlobal:
      return "G";
    case NT::DefVar:
      return "v";
    case NT::FunParam:
      return "p";
    case NT::RecField:
      return "f";
    case NT::ValString:
      return "S";
    case NT::ValCompound:
      return "C";
    default:
      return "";
  }
}

void LabelDefs(Node node, Prefix* prefix, std::map<Node, std::string>* labels) {
  std::string_view tag = GetLabelTag(node.kind());
  if (!tag.empty()) {
    (*labels)[node] = prefix->GetLabel(tag);
  }

  const NodeDesc& nd = GlobalNodeDescs[int(node.kind())];
  for (int i = 0; i < MAX_NODE_CHILDREN; ++i) {
    NFD_SLOT kind = nd.node_fields[i];
    Node child = Node_child_node(node, i);
    switch (GlobalNodeFieldDescs[int(kind)].kind) {
      case NFD_KIND::NODE:
        LabelDefs(child, prefix, labels);
        break;
      case NFD_KIND::LIST:
        prefix->PushLevel();
        for (int i = 0; !child.isnull(); ++i, child = Node_next(child)) {
          prefix->SetCurrent(i);
          LabelDefs(child, prefix, labels);
        }
        prefix->PopLevel();
        break;
      default:
        break;
    }
  }
}

void ExtractSymDefLabels(const std::vector<Node>& mods,
                         std::map<Node, std::string>* labels) {
  Prefix prefix;
  prefix.PushLevel();
  for (int i = 0; i < mods.size(); ++i) {
    Node mod = mods[i];
    prefix.SetCurrent(i);
    LabelDefs(mod, &prefix, labels);
  }
}

void ExtractTargetLabels(const std::vector<Node>& mods,
                         std::map<Node, std::string>* labels) {
  Prefix prefix;
  prefix.PushLevel();

  for (int i = 0; i < mods.size(); ++i) {
    Node mod = mods[i];
    prefix.SetCurrent(i);
    prefix.PushLevel();
    int block_counter = 0;
    int expr_counter = 0;

    auto visitor = [&labels, &prefix, &block_counter, &expr_counter](
                       Node node, Node parent) -> bool {
      switch (node.kind()) {
        case NT::StmtBlock:
          prefix.SetCurrent(block_counter++);
          (*labels)[node] = prefix.GetLabel("b");
          break;
        case NT::ExprStmt:
          prefix.SetCurrent(expr_counter++);
          (*labels)[node] = prefix.GetLabel("s");
        default:
          break;
      }
      return false;
    };

    VisitAstRecursivelyPre(mod, visitor, kNodeInvalid);
    prefix.PopLevel();
  }
}
constexpr int INDENTATION = 4;

void _EmitLine(const std::vector<std::string>& line, int indent,
               const std::vector<int>* active_columns, bool is_last) {
  if (indent >= 0) {
    std::vector<std::string_view> spaces(indent * INDENTATION + 1, " ");
    for (int pos : *active_columns) {
      spaces[pos * INDENTATION] = "┃";
    }
    spaces.back() = is_last ? "┗" : "┣";
    for (const auto& s : spaces) {
      std::cout << s;
    }
  }
  //
  std::string_view sep;
  for (const auto& s : line) {
    std::cout << sep << s;
    sep = " ";
  }
  std::cout << "\n";
}

// forward decl
void DumpNode(Node node, int indent, const std::map<Node, std::string>* labels,
              std::vector<int>* active_columns, bool is_last, bool dump_handles);

void DumpList(std::string_view name, Node node, int indent,
              const std::map<Node, std::string>* labels,
              std::vector<int>* active_columns, bool is_last, bool dump_handles) {
  std::vector<std::string> line = {std::string(name)};

  if (node.isnull()) {
    line.push_back("[Empty]");
    _EmitLine(line, indent + 1, active_columns, is_last);
  } else {
    _EmitLine(line, indent + 1, active_columns, is_last);
    active_columns->push_back(indent + 2);
    do {
      Node next = Node_next(node);
      if (next.isnull()) {
        active_columns->pop_back();
      }
      DumpNode(node, indent + 2, labels, active_columns, next.isnull(), dump_handles);
      node = next;
    } while (!node.isnull());
  }
}

int LastNodeSlot(const NodeDesc& desc) {
  int out = -1;
  for (int i = 0; i < MAX_NODE_CHILDREN; ++i) {
    NFD_SLOT slot = desc.node_fields[i];
    NFD_KIND kind = GlobalNodeFieldDescs[int(slot)].kind;
    if (kind == NFD_KIND::LIST || kind == NFD_KIND::NODE) {
      out = i;
    }
  }
  return out;
}

std::string RenderFlags(uint16_t compressed_flags, uint32_t desc_flag_bits) {
  if (compressed_flags == 0) {
    return "[]";
  }
  std::vector<std::string> flags;
  for (int i = 0; desc_flag_bits != 0; ++i, desc_flag_bits >>= 1) {
    if (desc_flag_bits & 1) {
      if (compressed_flags & Mask(BF(i))) {
        flags.push_back(std::string(EnumToString(BF(i))));
      }
    }
  }
  std::sort(flags.begin(), flags.end());
  std::string out;
  out.reserve(flags.size() * 15);  // basically we will stay below 16 only if
                                   // there is just on flag set
  std::string_view sep = "[";
  for (const auto& f : flags) {
    out += sep;
    out += f;
    sep = ", ";
  }
  out += "]";
  return out;
}

std::string RenderKind(Node node) {
  switch (node.kind()) {
    case NT::Expr1:
      return EnumToString(Node_unary_expr_kind(node));
    case NT::Expr2:
    case NT::StmtCompoundAssignment:
      return EnumToString(Node_binary_expr_kind(node));
    case NT::TypeBase:
      return EnumToString(Node_base_type_kind(node));
    case NT::ExprPointer:
      return EnumToString(Node_pointer_expr_kind(node));
    case NT::ModParam:
      return EnumToString(Node_mod_param_kind(node));
    case NT::MacroParam:
      return EnumToString(Node_macro_param_kind(node));
    case NT::DefMacro:
      return EnumToString(Node_macro_result_kind(node));
    case NT::DefEnum:
      return EnumToString(Node_base_type_kind(node));
    default:
      return "";
  }
}

void DumpNode(Node node, int indent, const std::map<Node, std::string>* labels,
              std::vector<int>* active_columns, bool is_last, bool dump_handles) {
  const NodeDesc& desc = GlobalNodeDescs[int(node.kind())];
  std::vector<std::string> line;

  auto add_tag_value = [&line](std::string_view tag, std::string_view value) {
    std::string out;
    out.reserve(tag.size() + value.size() + 1);
    out += tag;
    out += "=";
    out += value;
    line.push_back(out);
  };

  line.push_back(std::string(EnumToString(node.kind())));
  if (dump_handles) {
    line.push_back("[");
    line.push_back(std::to_string(node.index()));
    line.push_back("]");
  }
  if (desc.bool_field_bits != 0) {
    line.push_back(
        RenderFlags(Node_compressed_flags(node), desc.bool_field_bits));
  }

  if (Node_other_kind(node) != 0) {
    line.push_back(RenderKind(node));
  }

  for (int i = 0; i < MAX_NODE_CHILDREN; ++i) {
    NFD_SLOT slot = desc.node_fields[i];
    NFD_KIND kind = GlobalNodeFieldDescs[int(slot)].kind;
    switch (kind) {
      case NFD_KIND::STR: {
        Str str = Node_child_str(node, i);
        if (!str.isnull() && StrData(str)[0] != 0) {
          add_tag_value(EnumToString(slot), StrData(str));
        }
        break;
      }
      case NFD_KIND::NAME: {
        Name name = Node_child_name(node, i);
        if (!name.isnull()) {
          auto val = NameData(name);
          if (val[0] != 0) {
            if (slot == NFD_SLOT::name) {
              line.push_back(std::string(val));
            } else {
              add_tag_value(EnumToString(slot), val);
            }
          }
        }
        break;
      }

      default:
        break;
    }
  }

  if (labels->contains(node)) {
    add_tag_value("label", labels->at(node));
  }

  if (desc.has(NFD_X_FIELD::type)) {
    CanonType ct = Node_x_type(node);
    if (!ct.isnull()) {
      auto name = NameData(CanonType_name(ct));
      auto no = "[" + std::to_string(CanonType_get_original_typeid(ct)) + "]";
      if (strlen(name) < 16) {
        add_tag_value("x_type", std::string(name) + no);
      } else {
        add_tag_value("x_type", no);
      }
    }
  }

  if (desc.has(NFD_X_FIELD::eval)) {
    Const eval = Node_x_eval(node);
    if (!eval.isnull()) {
      add_tag_value("x_eval", to_string(eval, labels));
    }
  }

  if (desc.has(NFD_X_FIELD::symbol)) {
    Node def_sym = Node_x_symbol(node);
    ASSERT(labels->contains(def_sym), "" << Node_name(def_sym));
    if (!def_sym.isnull()) add_tag_value("x_symbol", labels->at(def_sym));
  }

  if (desc.has(NFD_X_FIELD::target)) {
    Node def_sym = Node_x_target(node);
    if (!def_sym.isnull()) add_tag_value("x_target", labels->at(def_sym));
  }

  if (desc.has(NFD_X_FIELD::poly_mod)) {
    Node def_sym = Node_x_poly_mod(node);
    if (!def_sym.isnull()) add_tag_value("x_poly_mod", labels->at(def_sym));
  }

  if (desc.has(NFD_X_FIELD::offset)) {
    add_tag_value("x_offset", std::to_string(Node_x_offset(node)));
  }

  _EmitLine(line, indent, active_columns, is_last);

  int last_slot = LastNodeSlot(desc);
  if (last_slot == -1) return;

  active_columns->push_back(indent + 1);
  for (int i = 0; i <= last_slot; ++i) {
    if (i == last_slot) {
      active_columns->pop_back();
    }
    NFD_SLOT slot = desc.node_fields[i];
    NFD_KIND kind = GlobalNodeFieldDescs[int(slot)].kind;
    Node child = Node_child_node(node, i);

    switch (kind) {
      case NFD_KIND::NODE:
        DumpNode(child, indent + 1, labels, active_columns, i == last_slot, dump_handles);
        break;
      case NFD_KIND::LIST:
        DumpList(std::string(EnumToString(slot)), child, indent, labels,
                 active_columns, i == last_slot, dump_handles);
        break;
      default:
        break;
    }
  }
}

void DumpAstMods(const std::vector<Node>& mods, bool dump_handles) {
  std::map<Node, std::string> labels;
  ExtractSymDefLabels(mods, &labels);
  ExtractTargetLabels(mods, &labels);
  std::vector<int> active_columns;
  for (Node mod : mods) {
    std::cout << "\n";
    active_columns.clear();
    DumpNode(mod, -1, &labels, &active_columns, true, dump_handles);
  }
}

}  // namespace cwerg::fe